aodndata/aatams_sattag_qc_dm/csv_schema.py

```diff
@@ -0,0 +1,279 @@
+
+        Args:
+          headers(list): a list of strings for keys
+          data_list(list): a list of strings with all data rows
+
+        Returns:
+          Dict[str,List[Any]]: A dict with header columns as keys
+          Dict[str,List[Any]]: A dict with header columns as keys
+          and all respective values in a list.
+
+        """
+        adict = {cname: [] for cname in headers}
+        for data in data_list:
+            for lind, key in enumerate(headers):
+                adict[key] += [data[lind]]
+        return adict
+
+    @staticmethod
+    def compute_cross_values(datadict, keys=None):
+        """Compute a cross value dictionary {k:set(v)}.
+
+        This dictionary is useful to compare
+        if a value of a key is within a group of values.
+
+        Args:
+          datadict(dict): the original dict.
+          keys():
+
+          keys: the selected keys.
+
+        Returns:
+
+        """
+        cross_values = {}
+        if set(keys).issubset(datadict.keys()):
+            cross_values = {x: set() for x in keys}
+            _ = [cross_values[x].update(datadict[x]) for x in cross_values.keys()]
+        return cross_values
+
+    @classmethod
+    def report(cls, msg):
+        """Report a msg.
+
+        Args:
+          msg(str): the msg
+
+        Returns:
+
+        """
+        print(cls.__name__ + ": " + msg)
+
+    def validate_file(self, file):
+        """Validate a csv file.
+
+        Args:
+          file(str): a csv file
+
+        Returns:
+          header(Dict[str,List[Any]]): A dict with header columns converted as keys
+          data(Dict[str,List[Any]]): A dict with data columns converted as keys
+
+        """
+        self.report("Validation Started for %s" % file)
+        self.report("\tLoading content schema for %s" % file)
+
+        schema_name = self.file2schema(file)
+        fschema = self.file_schemas[schema_name]
+
+        self.report("\tLoading data from %s" % file)
+        _, header, data_list = self.load_csv(
+            file, bulk_load=self.bulk_load, skip_every=self.skip_every,
+        )
+
+        self.report("\tValidating Header in %s" % file)
+        valid_header = self.validate_header(header, fschema)
+
+        self.report("\tAggregating data rows for validation in %s" % file)
+        col_dict = self.aggregate_to_dict(header, data_list)
+
+        self.report("\tValidating data types in %s" % file)
+        valid_dataset = {}
+        for key, value in col_dict.items():
+            # TODO continue to validate per column since the fail output can be verbose!?
+            valid_dataset[key] = Schema([fschema[key]]).validate(value)
+            self.report("\t\t %s[%s] is valid" % (file, key))
+
+        if self.same_keys_content:
+            self.report("\t\t\tValidating inter columns in %s" % file)
+            self.validate_self_content(self.same_keys_content, valid_dataset)
+
+        if self.cross_content_scope and self.cross_values:
```

Comment:
I'm trying hard to understand the logic with ```self.cross_values``` across the different functions, and running the unittests. It seems like the value is always empty, and the if block always bypassed. please enlight me!

Reply:
yeap - I will clarify with docstrings and fix what I have broken in the rebase.

It doesn't ring anything in tests because I changed the argument call to the function that computes the cross_values, which return empty results.

```cross_content_scope``` is a dictionary mapping a column name in one file to a different column name in another file. This is used by ```compute_cross_values``` to compare compare set(dict1[keyA]) against set(dict2[keyB]).  "cross"-> (among different dicts/files) "content" -> values , "scope" (set comparison) .

In the case of AATAMS, this is true for the "sattag_program", which is a field in metadata.csv, while this same field is named "cid" in other files.

My fail here is that cross_content_scope and related got no docs :)
 
If we change in line 673 in the aatams_schema:
```
-ckeys = self.cross_content_scope.keys()
+ckeys = self.cross_content_scope.values()
```
the cross content will be set-compared, but note that the naming is a bit of an anti-pattern ATM - a tuple is better .